File: FrontEnd/lib/domain/usecases/reminders/get_user_reminders_use_case.dart
/// @file get_user_reminders_use_case.dart
/// @description Implementación del caso de uso para obtener los recordatorios del usuario.
/// @module Reminders
/// @layer Domain
library;

import '../../entities/reminder.dart';
import '../../interfaces/usecases/reminders/i_get_user_reminders_use_case.dart';
import '../../repositories/i_reminder_repository.dart';

/// [implements] IGetUserRemindersUseCase
/// [dependencies] IReminderRepository
class GetUserRemindersUseCase implements IGetUserRemindersUseCase {
  final IReminderRepository _repository;
  const GetUserRemindersUseCase({required IReminderRepository repository})
      : _repository = repository;

  @override
  Future<List<Reminder>> execute() => _repository.getActiveReminders();
}

File: BackEnd/src/domain/usecases/reminders/GetUserRemindersUseCase.ts
/**
 * @file GetUserRemindersUseCase.ts
 * @description Caso de uso para obtener los recordatorios activos de un usuario.
 * @module Reminders
 * @layer Domain
 *
 * @implements {IGetUserRemindersUseCase}
 * @injectable
 * @dependencies IReminderRepository, IReminderMapper
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IGetUserRemindersUseCase } from '../../interfaces/usecases/reminders/IGetUserRemindersUseCase.js';
import type { IReminderRepository } from '../../repositories/IReminderRepository.js';
import type { IReminderMapper } from '../../../data/IMappers/IReminderMapper.js';
import type { ReminderResponseDTO } from '../../dtos/reminders/reminder-response.dto.js';

/**
 * Obtiene los recordatorios activos (no completados) del usuario.
 *
 * @implements {IGetUserRemindersUseCase}
 * @injectable
 * @dependencies IReminderRepository, IReminderMapper
 */
@injectable()
export class GetUserRemindersUseCase implements IGetUserRemindersUseCase {
  constructor(
    @inject(TYPES.IReminderRepository) private readonly reminderRepo: IReminderRepository,
    @inject(TYPES.IReminderMapper)     private readonly mapper: IReminderMapper,
  ) {}

  /**
   * @param userId — Id del usuario autenticado.
   * @returns Lista de ReminderResponseDTO ordenados por scheduledDate asc.
   */
  async execute(userId: string): Promise<ReminderResponseDTO[]> {
    const reminders = await this.reminderRepo.findByUserId(userId);
    return reminders.map((r) => this.mapper.toResponseDTO(r));
  }
}
